README
NAME
    lib::abs - "lib" that makes relative path absolute to caller.

SYNOPSIS
    Simple use like "use lib ...":

        use lib::abs qw(./mylibs1 ../mylibs2);
        use lib::abs 'mylibs';

    Extended syntax (glob)

        use lib::abs 'modules/*/lib';

    There are also may be used helper function from lib::abs (see
    example/ex4):

        use lib::abs;
        # ...
        my $path = lib::abs::path('../path/relative/to/me'); # returns absolute path

DESCRIPTION
    The main reason of this library is transformate relative paths to
    absolute at the "BEGIN" stage, and push transformed to @INC. Relative
    path basis is not the current working directory, but the location of
    file, where the statement is (caller file). When using common "lib",
    relative paths stays relative to curernt working directory,

        # For ex:
        # script: /opt/scripts/my.pl
        use lib::abs '../lib';

        # We run `/opt/scripts/my.pl` having cwd /home/mons
        # The @INC will contain '/opt/lib';

        # We run `./my.pl` having cwd /opt
        # The @INC will contain '/opt/lib';

        # We run `../my.pl` having cwd /opt/lib
        # The @INC will contain '/opt/lib';

    Also this module is useful when writing tests, when you want to load
    strictly the module from ../lib, respecting the test file.

        # t/00-test.t
        use lib::abs '../lib';

    Also this is useful, when you running under "mod_perl", use something
    like "Apache::StatINC", and your application may change working
    directory. So in case of chdir "StatINC" fails to reload module if the
    @INC contain relative paths.

RATIONALE
    Q: We already have "FindBin" and "lib", why we need this module?

    A: There are several reasons:

    1) "FindBin" could find path incorrectly under "mod_perl"
    2) "FindBin" works relatively to executed binary instead of relatively
    to caller
    3) Perl is linguistic language, and `use lib::abs "..."' semantically
    more clear and looks more beautiful than `use FindBin; use lib
    "$FindBin::Bin/../lib";'
    4) "FindBin" b<will> work incorrectly, if will be called not from
    executed binary (see <http://github.com/Mons/lib-abs-vs-findbin>
    comparison for details)

BUGS
    None known

COPYRIGHT & LICENSE
    Copyright 2007-2010 Mons Anderson.

    This program is free software; you can redistribute it and/or modify it
    under the same terms as Perl itself.

AUTHOR
    Mons Anderson, "<mons@cpan.org>"

CONTRIBUTORS
    Oleg Kostyuk, "<cub@cpan.org>"



Changes
Revision history for lib-abs

0.92    2010-11-15
        Added "WHY" section into POD

0.91    2010-02-27
        Fix broken test due to behavior of lib under 5.10.1
        Fix broken test (warn) when hit some file as directory

0.90    2009-07-21
        Renamed ex-lib to lib abs
        Make import args to use glob patterns

Revision history for ex-lib

0.07    2009-05-19
        Added lib::abs initially

0.06    2009-04-20
        Allow use with empty list
        Made kwalitee.t to check $ENV{TEST_AUTHOR}
        
0.05    2009-03-12
        Lower dependency on Cwd from 3.29 to 3.12
        Add tests for different versions of Cwd
        Fix $@ bug

0.04    2009-01-29
        Fixes for win32
        Add strict dependency on Cwd

0.03    2009-01-28
        Fixes for perl < 5.8

0.02    2009-01-23
        Minor fixes

0.01    2008-09-09
        First version, released on an unsuspecting world.



t/check.pl
use strict;

my $tests; BEGIN { $tests = 22; }

use Test::More;

use FindBin;
BEGIN {
	defined $FindBin::Bin and length $FindBin::Bin
		or plan skip_all => "\$FindBin::Bin (v$FindBin::VERSION) is ".(defined $FindBin::Bin ? 'empty' : 'undef').". Not my fault";
}

use Carp ();
use Carp::Heavy (); # lib::abs may require Carp in runtime. With modified INC it may fail
use overload ();    # Test::More uses overload in runtime. With modified INC it may fail

use lib '.',"$FindBin::Bin/../lib",;
BEGIN { eval { require Test::NoWarnings; Test::NoWarnings->import; 1 } and ++$tests }

my @ORIG; BEGIN { @ORIG = @INC }

sub ex { @INC[0..@INC-@ORIG-1] }
sub diaginc { 0 or return; diag +( @_ ? ($_[0].': ') : ( 'Add INC: ') ) . join ', ', map "'$_'", ex(); }

#BEGIN { *lib::abs::DEBUG = sub () { 2 } }
use lib::abs ();

diag( "Testing lib::abs $lib::abs::VERSION using Cwd $Cwd::VERSION with FindBin $FindBin::VERSION, Perl $], $^X" );
diaginc();

plan tests => $tests;

is( $INC[0], ".", "before import: $INC[0]" );
lib::abs->import( '.' );
diag "Bin = `$FindBin::Bin' ;. is `$INC[0]'";
is( $FindBin::Bin, $INC[0], '. => $FindBin::Bin' );

diaginc();

lib::abs->unimport( '.' );
ok(!ex, 'no ex inc');

diaginc();

# Next tests are derived from lib::tiny


my @dirs = qw(foo bar);
my @adirs = map "$FindBin::Bin/$_",@dirs;
#printf "%o\n", umask(0);
mkdir($_, 0755) or warn "mkdir $_: $!" for @adirs;
chmod 0755, @adirs or warn "chmod $_: $!"; # do chmod (on some versions mkdir with mode ignore mode)

-e $_ or warn "$_ absent" for @adirs;

lib::abs->import(@dirs);

diaginc();
is($INC[0],$adirs[0],'add 0');
is($INC[1],$adirs[1],'add 1');

lib::abs->unimport(@dirs);
diaginc();

ok(!ex, 'dels paths');

eval {
    require lib;
    'lib'->import(@adirs);
};

SKIP: {
    skip 'apparently too old to handle: Unquoted string "lib" may clash with future reserved word at t/00.load.t line 21.', 1 if $@;
	is($INC[0],$adirs[0],'order same as lib.pm 0');
	is($INC[1],$adirs[1],'order same as lib.pm 1');
};

eval {
    'lib'->unimport(@adirs);
};

lib::abs->import( '.' );

# Reflib

@INC = ();
lib::abs->import( sub {} );
my $chk = shift @INC; # When left bad sub in @INC Test::Builder fails
is(ref $chk, 'CODE', 'code in @INC');

# Abs ok

@INC = ();
lib::abs->import(
	'///opt/perl/lib',
	'//opt/perl/lib',
	'/opt/perl/lib',
);
my @abs = @INC; @INC = ();
lib->import(
	'///opt/perl/lib',
	'//opt/perl/lib',
	'/opt/perl/lib',
);
my @need = @INC; @INC = ();
is_deeply \@abs,\@need, 'absolute is same as lib';

# Rel ok

@INC = ();
lib::abs->import(
	'.///',
	'.//',
	'./',
	'.',
);
my @chk = @INC; @INC = ();

SKIP: {
	is($chk[0], $FindBin::Bin,     './// => .');
	@chk > 1 or skip "Duplicates are collapsed",3;
	is($chk[1], $FindBin::Bin,     '.// => .');
	is($chk[2], $FindBin::Bin,     './ => .');
	is($chk[3], $FindBin::Bin,     '. => .');
}

# Glob test

lib::abs->import('glob/*/lib');
@chk = @INC; @INC = ();
my @have;
for (@chk) {
	my $i = m{/glob/t(\d)/lib$} ? $1 : 99;
	$have[$i] = 1;
	like($_, qr{/glob/t(\d)/lib$}, 'glob t'.$i);
	
	
}
is_deeply( \@have,[1,1,1], 'all items present' );

lib::abs->import('glob/x?x/inc');
@chk = @INC; @INC = ();
@have = ();
for (@chk) {
	my $i = m{/glob/x(.)x/inc$} ? $1 : 99;
	$have[$i] = 1;
	like($_, qr{/glob/x(.)x/inc$}, 'glob x'.$i.'x');
}
is_deeply( \@have,[1,1], 'all items present' );

is ( lib::abs::path('.'), $FindBin::Bin, 'lib::abs::path' );
diag "lib::abs::path('.')=".lib::abs::path('.');

exit 0;

END{
	rmdir $_ for @adirs; # clean up
}


lib/lib/abs.pm
#
# Copyright (c) 2007-2010 Mons Anderson <mons@cpan.org>. All rights reserved
# This program is free software; you can redistribute it and/or
# modify it under the same terms as Perl itself.
package lib::abs;

=head1 NAME

lib::abs - C<lib> that makes relative path absolute to caller.

=cut

our $VERSION = '0.92';

=head1 SYNOPSIS

Simple use like C<use lib ...>:

	use lib::abs qw(./mylibs1 ../mylibs2);
	use lib::abs 'mylibs';

Extended syntax (glob)

	use lib::abs 'modules/*/lib';

There are also may be used helper function from lib::abs (see example/ex4):

	use lib::abs;
	# ...
	my $path = lib::abs::path('../path/relative/to/me'); # returns absolute path

=head1 DESCRIPTION

The main reason of this library is transformate relative paths to absolute at the C<BEGIN> stage, and push transformed to C<@INC>.
Relative path basis is not the current working directory, but the location of file, where the statement is (caller file).
When using common C<lib>, relative paths stays relative to curernt working directory, 

	# For ex:
	# script: /opt/scripts/my.pl
	use lib::abs '../lib';

	# We run `/opt/scripts/my.pl` having cwd /home/mons
	# The @INC will contain '/opt/lib';

	# We run `./my.pl` having cwd /opt
	# The @INC will contain '/opt/lib';

	# We run `../my.pl` having cwd /opt/lib
	# The @INC will contain '/opt/lib';

Also this module is useful when writing tests, when you want to load strictly the module from ../lib, respecting the test file.

	# t/00-test.t
	use lib::abs '../lib';

Also this is useful, when you running under C<mod_perl>, use something like C<Apache::StatINC>, and your application may change working directory.
So in case of chdir C<StatINC> fails to reload module if the @INC contain relative paths.

=head1 RATIONALE

Q: We already have C<FindBin> and C<lib>, why we need this module?

A: There are several reasons:

=over 4

=item 1) C<FindBin> could find path incorrectly under C<mod_perl>

=item 2) C<FindBin> works relatively to executed binary instead of relatively to caller

=item 3) Perl is linguistic language, and C<`use lib::abs "..."'> semantically more clear and looks more beautiful than C<`use FindBin; use lib "$FindBin::Bin/../lib";'>

=item 4) C<FindBin> b<will> work incorrectly, if will be called not from executed binary (see L<http://github.com/Mons/lib-abs-vs-findbin> comparison for details)

=back

=head1 BUGS

None known

=head1 COPYRIGHT & LICENSE

Copyright 2007-2010 Mons Anderson.

This program is free software; you can redistribute it and/or modify it
under the same terms as Perl itself.

=head1 AUTHOR

Mons Anderson, C<< <mons@cpan.org> >>

=head1 CONTRIBUTORS

Oleg Kostyuk, C<< <cub@cpan.org> >>

=cut

use 5.006;
use strict;
use warnings;
use lib ();
use Cwd 3.12 qw(abs_path);
$lib::abs::sep = {
	( map { $_ => qr{[^\\/]+$}o } qw(mswin32 netware symbian dos) ),
	( map { $_ => qr{[^:]+:?$}o } qw(macos) ),
}->{lc$^O} || qr{[^/]+$}o;

BEGIN { *DEBUG = sub () { 0 } unless defined &DEBUG } # use constants is heavy

sub _carp  { require Carp; goto &Carp::carp  }
sub _croak { require Carp; goto &Carp::croak }
sub _debug ($@) { printf STDERR shift()." at @{[ (caller)[1,2] ]}\n",@_ }

sub mkapath($) {
	my $depth = shift;
	
	# Prepare absolute base bath
	my ($pkg,$file) = (caller($depth))[0,1];
	_debug "file = $file " if DEBUG > 1;
	$file =~ s/${lib::abs::sep}//s;
	$file = '.' unless length $file;
	_debug "base path = $file" if DEBUG > 1;
	my $f = abs_path($file) . '/';
	_debug "source dir = $f " if DEBUG > 1;
	$f;
}

sub path {
	local $_ = shift;
	s{^\./+}{};
	local $!;
	my $abs = mkapath(1) . $_;
	my $ret = abs_path( $abs ) or _carp("Bad path specification: `$_' => `$abs'" . ($! ? " ($!)" : ''));
	_debug "$_ => $ret" if DEBUG > 1;
	$ret;
}

sub transform {
	my $prefix;
	map {
		ref || m{^/} ? $_ : do {
			my $lib = $_;
			s{^\./+}{};
			local $!;
			my $abs = ( $prefix ||= mkapath(2) ) . $_;
			if (index($abs,'*') != -1 or index($abs,'?') !=-1) {
				_debug "transforming $abs using glob" if DEBUG > 1;
				map {
					abs_path( $_ )
						or _croak("Bad path specification: `$lib' => `$_'" . ($! ? " ($!)" : ''))
				} glob $abs;
			} else {
				$_ = abs_path( $abs ) or _croak("Bad path specification: `$lib' => `$abs'" . ($! ? " ($!)" : ''));
				_debug "$lib => $_" if DEBUG > 1;
				($_);
			}
		}
	} @_;
}

sub import {
	shift;
	return unless @_;
	@_ = ( lib => transform @_ = @_ );
	_debug "use @_\n" if DEBUG > 0;
	goto &lib::import;
	return;
}

sub unimport {
	shift;
	return unless @_;
	@_ = ( lib => transform @_ = @_ );
	_debug "no @_\n" if DEBUG > 0;
	goto &lib::unimport;
	return;
}

1;
